### ISSUE TITLE
KeyError crash issue near Y2 rock.

### ISSUE THREAD

**Issue by mikesiegel** [brandon-rhodes/python-adventure#26]
```
> n
YOU ARE IN A LOW N/S PASSAGE AT A HOLE IN THE FLOOR.  THE HOLE GOES
DOWN TO AN E/W PASSAGE.

THERE IS A SMALL WICKER CAGE DISCARDED NEARBY.

> n
Traceback (most recent call last):
  File "/usr/games/adventure", line 9, in <module>
    load_entry_point('adventure==1.4', 'console_scripts', 'colossal-cave-adventure')()
  File "/usr/share/colossal-cave-adventure/adventure/__main__.py", line 101, in main
    loop(game)
  File "/usr/share/colossal-cave-adventure/adventure/__main__.py", line 79, in loop
    baudout(game.do_command(words))
  File "/usr/share/colossal-cave-adventure/adventure/game.py", line 437, in do_command
    self._do_command(words)
  File "/usr/share/colossal-cave-adventure/adventure/game.py", line 506, in _do_command
    self.dispatch_command(words)
  File "/usr/share/colossal-cave-adventure/adventure/game.py", line 553, in dispatch_command
    return self.do_motion(word1)
  File "/usr/share/colossal-cave-adventure/adventure/game.py", line 701, in do_motion
    self.move_to(move.action)
  File "/usr/share/colossal-cave-adventure/adventure/game.py", line 186, in move_to
    self.move_dwarves()
  File "/usr/share/colossal-cave-adventure/adventure/game.py", line 325, in move_dwarves
    self.describe_location()
  File "/usr/share/colossal-cave-adventure/adventure/game.py", line 379, in describe_location
    self.write(obj.messages[prop])
KeyError: 0
```

Happened on stream at 23 minutes if you want to check it out too.  I'll see if I can figure out how to reproduce it. 
https://www.twitch.tv/videos/709451376

**Comment by brandon-rhodes** [comment]
During a stream!? Oh, no! I'll go take a look and see if I can guess what happened.

**Comment by brandon-rhodes** [comment]
(I edited your comment to switch the triple-single-quotes to triple-backticks, which I think is what was intended.)

**Comment by brandon-rhodes** [comment]
I’ll have to leave the keyboard in a moment here and investigate later, but: it looks like each object has a dict of messages that can be printed to describe the object sitting in a room. For example (here playing Adventure at the Python prompt):
```
import adventure
>>> adventure._game.objects['lamp'].messages
{0.0: 'THERE IS A SHINY BRASS LAMP NEARBY.\n', 1.0: 'THERE IS A LAMP SHINING NEARBY.\n'}
```
The error is about an object that had no `messages` entry for the number zero. What objects have that property (assuming that messages are never created or deleted, which I think is true)? Let's look:
```
>>> objects = {obj for obj in adventure._game.objects.values() if 0 not in obj.messages}
>>> for o in objects: print(o)
... 
<Object 17 dwarf/dwarves 7f4357076ac0>
<Object 18 knife/knives 7f4357076be0>
<Object 21 water/h2o 7f4357076f40>
<Object 22 oil 7f435707b0a0>
```
So presumably one of those four objects wound up located at your position — which should have been at Y2, I think, if you went north from that location? — and when the game tried to display its description from the `messages` array, it died.

So I wonder which object it was. And I wonder why dwarves don't have messages attached? It would be an easy way to get their description up on the screen.

I should have time to look more later or this weekend. Thanks for letting me know!

**Comment by mikesiegel** [comment]
Thanks for jumping in so quick, I think that's the quickest response to an issue I've ever seen!

 It occurs to me those 4 objects are present in rooms, but not visible on the 'ground' as far as I recall. You can't see the dwarf standing there, or the thrown knife.

**Comment by brandon-rhodes** [comment]
1. I have never received a bug report before accompanied by a stream. It’s amazingly helpful! Unlike normal bug reports, which lack context, I can see that this happened after a death and reincarnation and after you returned to the same place where you had just died. An important clue!
2. What’s the soundtrack? I like the pipes.
3. Alas that this happened during a stream, though! A whole audience will now avoid Adventure because it crashes with a Python traceback.
4. Is that *the* CowboyNeal?
5. And, to answer your comment on the stream: I’m not sure why, when the original Fortran version of Adventure is freely available and could be compiled and run, they decided to instead use my hand-translated Python version that I mostly wrote to practice Python 3 (but also to learn any secrets or Easter eggs in the original code). I do like my simulated baud rate, which makes the game feel more like the terminals I first played it on; it adds suspense to wait for the messages to finish printing.
6. So I played through a bit myself to create the same situation that you encountered, recording my terminal in the meantime, and having (a) manually set the random number seed at the beginning so that the playthrough could be reproduced exactly and (b) added a more informative exception. And it worked! On the first try! The message that printed out was:
```
    ValueError: <Object 21 water/h2o 7f81c3c20e80> [<room 33 at 0x7f81c38944c0>]
```
What? The problem is the water?

Let’s go look at that death code again.

```
                    # do water and oil thing
```

OH NO! It’s all my fault!

Apparently I was planning to go back and finish that and didn’t. These days I use the keyword `TODO` in such comments but, apparently, I wasn’t yet consistent about that habit back in December of 2010. So that comment is still sitting there, and the oil-and-water case is not yet properly handled by the code.

The code drops all of your currently-carried objects on the floor. Except for oil and water, because they can’t sit on the floor separately from a bottle. So the original Fortran says:

```
	PLACE(WATER)=0
	PLACE(OIL)=0
```

I will now go do the equivalent in the Python code, as I ought to have done when I originally wrote it, and see if that fixes it.

**Comment by brandon-rhodes** [comment]
@mikesiegel — The three commits above (a) solve the problem, (b) keep it from breaking a new circumstance (when someone dies with an empty bottle), and (c) solves a longstanding problem I had never noticed: that when you drop a full bottle, its description still say it’s empty.

The latter problem was because I apparently overlooked the Fortran line:
```
	PROP(BOTTLE)=MOD(COND(LOC),4)/2*2
```

If you want to install it locally to try out the fixes:
```
pip install https://github.com/brandon-rhodes/python-adventure/archive/master.zip
```

But since you might have installed it from your OS, I'll understand if you don't want to bother trying it out — especially since dying with a bottle full of water is an annoying circumstance to have to arrange!

I might also look into your complaint during the stream that all-caps commands are not accepted. That seems like it should be fixed!

Thanks for the report, it would have been easy to not bother with, but fixed a problem that might have been affecting folks for years without my ever hearing of it!

**Comment by mikesiegel** [comment]
Great, thanks for the quick fix. Installed and next time I play adventure I'll do my best to die with a full bottle. And yes, that is the CowboyNeal. Music by Corvus Covax probably, perhaps 'Twilight of the Thunder Gods'. A playlist I put together for MUDing.

**Comment by brandon-rhodes** [comment]
1. I also went in and added support for commands typed at the terminal in upper case, to resolve another issue I noted in your livestream; I'm not sure why it hadn't ever occurred to me to make the input case-insensitive, given that the game is always typing at folks in all uppercase.
2. I have just released 1.6 featuring this fix. Hopefully that makes its way at some point into distributions!
